### PR TITLE
drivers: sdhc: add driver support for emmc host controller

### PR DESCRIPTION
**Pull request by najumon1980**
add driver support for emmc host controller (eMMC 5.1 compliant ). Following are some of the key supported features:
- Supported Mode: HS400, HS200, DDR50, SDR104, SDR50, SDR25 and SDR12
- DMA support: Both SDMA and ADMA supported
- Data Width: 1bit, 4bit and 8Bit
- Support Interrupt driven as well as polled mode

### CODE REVIEW

File: drivers/sdhc/emmc_host.c [hunk 1]
```diff
@@ -0,0 +1,1337 @@
...
+			ret = emmc_send_cmd_data(emmc, MMC_SEND_EXT_CSD, cmd, data, true);
+			break;
+
+		default:
+			ret = emmc_send_cmd_data(emmc, cmd->opcode, cmd, data, true);
+		}
+	} else {
+		ret = emmc_send_cmd_no_data(emmc, cmd->opcode, cmd);
+
+		/*fixme: WA for adding delay after card switch to HS200 mode */
```
Comment (danieldegrasse) [inline drivers/sdhc/emmc_host.c:986]: Just out of curiosity- what is the specific behavior this works around? It seems like this line is intended to induce a delay when calling `sdmmc_wait_ready` here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/sd/mmc.c#L454

I wonder a few things here:
- should `sdhc_busy` be returning true after the card switches to HS200 mode until the 10 ms delay expires? This way, the `SD_SEND_STATUS` command is not sent until the card is ready for data
- would increasing `CONFIG_SD_DATA_TIMEOUT` work around this issue as well? IE is the card timing out during the call to `sdmmc_wait_ready`?

File: drivers/sdhc/emmc_host.c [hunk 2]
```diff
@@ -0,0 +1,1337 @@
...
+	/* Set HC signal voltage */
+	if (ios->signal_voltage != host_io->signal_voltage) {
+		LOG_DBG("signal_voltage: %d", ios->signal_voltage);
+		ret = emmc_set_voltage(emmc, ios->signal_voltage);
+		if (ret) {
+			LOG_ERR("Set signal volatge failed failed:%d", ret);
+			return ret;
+		}
+		host_io->signal_voltage = ios->signal_voltage;
+		k_msleep(10u);
```
Comment (danieldegrasse) [inline drivers/sdhc/emmc_host.c:1045]: What is the reason for the delay here? I assume there is some requirement in the datasheet, but would like a comment to that effect here

File: drivers/sdhc/emmc_host.c [hunk 3]
```diff
@@ -0,0 +1,1337 @@
...
+	}
+
+	/* Set card power */
+	if (host_io->power_mode != ios->power_mode) {
+		LOG_DBG("power_mode: %d", ios->power_mode);
+
+		ret = emmc_set_power(emmc, ios->power_mode);
+		if (ret) {
+			LOG_ERR("Set Bus Width failed:%d", ret);
```
Comment (danieldegrasse) [inline drivers/sdhc/emmc_host.c:1054]: Error log here needs to be updated

File: drivers/sdhc/emmc_host.c [hunk 4]
```diff
@@ -0,0 +1,1337 @@
...
+	return ((int)emmc->card_present);
+}
+
+static int emmc_execute_tuning(const struct device *dev)
+{
+#if defined(CONFIG_EMMC_HOST_TUNING)
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+
+	LOG_ERR("Tuning starting...");
```
Comment (danieldegrasse) [inline drivers/sdhc/emmc_host.c:1097]: Why use `LOG_ERR` here? does the user need this log any time they enable basic SDMMC logging?

File: drivers/sdhc/emmc_host.c [hunk 5]
```diff
@@ -0,0 +1,1337 @@
...
+
+	regs->host_ctrl2 |= EMMC_HOST_START_TUNING;
+	while (!(regs->host_ctrl2 & EMMC_HOST_START_TUNING)) {
+		;
+	}
+
+	if (regs->host_ctrl2 & EMMC_HOST_TUNING_SUCCESS) {
+		LOG_ERR("Tuning Completed success");
```
Comment (danieldegrasse) [inline drivers/sdhc/emmc_host.c:1105]: Same question as above on the log level here

File: drivers/sdhc/emmc_host.c [hunk 6]
```diff
@@ -0,0 +1,1337 @@
...
+	props->f_min = config->min_bus_freq;
+	props->power_delay = config->power_delay_ms;
+
+	props->host_caps.vol_180_support = (bool)(cap & BIT(26u));
+	props->host_caps.vol_300_support = (bool)(cap & BIT(25u));
+	props->host_caps.vol_330_support = (bool)(bool)(cap & BIT(24u));
+	props->host_caps.suspend_res_support = false;
+	props->host_caps.sdma_support = (bool)(cap & BIT(22u));
+	props->host_caps.high_spd_support = (bool)(cap & BIT(21u));
+#if defined(CONFIG_EMMC_HOST_ADMA)
```
Comment (danieldegrasse) [inline drivers/sdhc/emmc_host.c:1148]: Could this be a DT property of the hardware? It seems like this would be a property of the IP itself, not a software configuration. Alternatively, could this be probed at runtime since the driver uses PCIE?

File: drivers/sdhc/Kconfig.emmc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	depends on EMMC_HOST_DMA
+	help
+	  Block gap mode support.
+
+config EMMC_HOST_TUNING
+	bool "EMMC re-tuning support"
+	help
+	  EMMC re-tuing support.
+
+module = EMMC_HOST
```
Comment (danieldegrasse) [inline drivers/sdhc/Kconfig.emmc:56]: Why are additional logging Kconfigs being added here? `CONFIG_SDHC_LOG_LEVEL` should be used to set the log level in this driver

File: dts/bindings/sdhc/sdhc.yaml [hunk 8]
```diff
@@ -59,3 +59,13 @@ properties:
     type: boolean
     description: |
       The host controller supports HS400 at 1.8V
+
+  mmc-dw-8bit:
```
Comment (danieldegrasse) [inline dts/bindings/sdhc/sdhc.yaml:63]: This commit needs to be first in the PR, since the emmc driver depends on these properties.

Also, please use a title similar to the following:

`dts: bindings: sdhc: add dts entry for sdhc bus width`

File: subsys/sd/sd.c [hunk 9]
```diff
@@ -117,17 +117,32 @@ static int sd_common_init(struct sd_card *card)
...
+	int ret, voltage;
 
 	/* SD clock should start gated */
 	bus_io->clock = 0;
 	/* SPI requires SDHC PUSH PULL, and open drain buses use more power */
 	bus_io->bus_mode = SDHC_BUSMODE_PUSHPULL;
 	bus_io->power_mode = SDHC_POWER_ON;
 	bus_io->bus_width = SDHC_BUS_WIDTH1BIT;
-	/* Cards start with legacy timing and 3.3V signalling at power on */
+	/* Cards start with legacy timing and Maximum volatge HC support */
```
Comment (danieldegrasse) [inline subsys/sd/sd.c:129]: Nit- spelling

File: subsys/sd/sd.c [hunk 10]
```diff
@@ -117,17 +117,32 @@ static int sd_common_init(struct sd_card *card)
...
+		voltage = SD_VOL_3_3_V;
+	} else if (host_props->host_caps.vol_300_support) {
+		LOG_DBG("HC support 3.0V max");
+		voltage = SD_VOL_3_0_V;
+	} else {
+		LOG_DBG("HC support 1.8V max");
+		voltage = SD_VOL_1_8_V;
+	}
+
+	/* Set to maximum voltage support by HC. */
```
Comment (danieldegrasse) [inline subsys/sd/sd.c:143]: Nit- please spell out "host controller" here, it makes the comment clearer IMO

File: drivers/sdhc/emmc_host.c (hunk 2 above)
Comment (najumon1980) [reply drivers/sdhc/emmc_host.c]: This delay added as a safe point since power switching happens (not a requirement as per HC spec , section 3.3 SD Bus Power control). I removed this delay to avoid unnecessary confusion.

File: drivers/sdhc/emmc_host.c (hunk 6 above)
Comment (najumon1980) [reply drivers/sdhc/emmc_host.c:1142]: Make it DTS config is a good suggestion, but too many parameter via DTS was my concern.

Comment (danieldegrasse) [reply drivers/sdhc/emmc_host.c:1142]: I'd prefer this to be a DTS setting over Kconfig

File: subsys/sd/sd.c [hunk 11]
```diff
@@ -144,8 +159,8 @@ static int sd_init_io(struct sd_card *card)
 		LOG_ERR("Could not disable card power via SDHC");
 		return ret;
 	}
-	/* After reset or init, card voltage should be 3.3V */
-	card->card_voltage = SD_VOL_3_3_V;
+	/* After reset or init, card voltage should be max HC support */
```
Comment (decsny) [inline subsys/sd/sd.c:162]: what is the reason for this change?

File: dts/bindings/sdhc/sdhc.yaml [hunk 12]
```diff
@@ -59,3 +59,13 @@ properties:
...
+    type: boolean
+    description: |
+      The host controller supports 8bit data width
+
+  mmc-dw-4bit:
+    type: boolean
```
Comment (decsny) [inline dts/bindings/sdhc/sdhc.yaml:69]: do these need to be two separate properties? and does it need to be specific to mmc?

File: dts/bindings/sdhc/emmc-host.yaml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+description: eMMC host controller
+
+compatible: "emmc-host"
```
Comment (decsny) [inline dts/bindings/sdhc/emmc-host.yaml:6]: this naming is too generic, it is confusing, it should at least be prefixed with the vendor name like `intel,emmc-host`

**Review by decsny — DISMISSED** [review]
requesting change to compatible string

File: dts/bindings/sdhc/emmc-host.yaml (hunk 13 above)
Comment (najumon1980) [reply dts/bindings/sdhc/emmc-host.yaml:6]: the driver is compline to the standard host specification and shall work with the most of the EMMC host controller IP.  That is why I given it a generic eMMC Host compatibility string.

File: dts/bindings/sdhc/sdhc.yaml (hunk 8 above)
Comment (najumon1980) [reply dts/bindings/sdhc/sdhc.yaml:63]: This commit is before driver commit.

File: dts/bindings/sdhc/sdhc.yaml (hunk 12 above)
Comment (najumon1980) [reply dts/bindings/sdhc/sdhc.yaml:69]: When Driver change to higher bus speed mode (HS200, HS400 etc.) , the bus width also need to change and based on host capabilities (whether it support 8bit or not etc.) it can decide whether it can change higher speed. So it need to be separate.

File: subsys/sd/sd.c (hunk 11 above)
Comment (najumon1980) [reply subsys/sd/sd.c:162]: eMMC host controller might support maximum 1.8V and hence change to accommodate setting voltage based on HC capabilities instead of hard coding to 3.3V.

File: dts/bindings/sdhc/sdhc.yaml (hunk 12 above)
Comment (decsny) [reply dts/bindings/sdhc/sdhc.yaml:69]: okay, what I meant is would it better to combine them to a single property that was a description of the bus width like linux does

File: dts/bindings/sdhc/emmc-host.yaml (hunk 13 above)
Comment (danieldegrasse) [reply dts/bindings/sdhc/emmc-host.yaml:6]: What devices has this driver been tested on? If we can verify this driver functions on multiple host controller IP blocks, I am ok with a more generic binding name. I suspect we will need to remove (or conventionalize) the PCIE macros to do this though.

If we want to introduce this as a generic driver though, I would prefer we expose the SD host controller function implementations in a header file similar to how Linux does, or create an `sdhc_common.c` file that this driver (and others) can utilize). That way vendors can leverage the generic SDHC functions where possible in their driver implementations, and override the SDHC functions they need to.

File: drivers/sdhc/emmc_host.c (hunk 6 above)
Comment (najumon1980) [reply drivers/sdhc/emmc_host.c]: updated to read from capabilities register.

File: dts/bindings/sdhc/sdhc.yaml (hunk 12 above)
Comment (najumon1980) [reply dts/bindings/sdhc/sdhc.yaml:69]: I am not sure how or whether we can use it to determine the specific capabilities supported. 

https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/mmc/mmc-controller.yaml
  bus-width:
    description:
      Number of data lines.
    $ref: /schemas/types.yaml#/definitions/uint32
    enum: [1, 4, 8]
    default: 1

File: dts/bindings/sdhc/emmc-host.yaml (hunk 13 above)
Comment (najumon1980) [reply dts/bindings/sdhc/emmc-host.yaml:6]: Currently I will change the compatibility id to "intel,emmc-host" since I tested only with Intel platform. But later someone can update as per your suggested once they tested with different platform.

File: dts/bindings/sdhc/sdhc.yaml (hunk 12 above)
Comment (decsny) [reply dts/bindings/sdhc/sdhc.yaml:69]: why not? does it not work for linux? If I remember correctly, the rest of the binding properties we tried to align to linux

File: drivers/sdhc/emmc_host.c [hunk 14]
```diff
@@ -0,0 +1,1328 @@
...
+	/* Keep enable interrupt status register to update */
+	regs->normal_int_stat_en = EMMC_HOST_NORMAL_INTR_MASK;
+	regs->err_int_stat_en = EMMC_HOST_ERROR_INTR_MASK;
+
+	/* Disable only interrupt generation */
+	regs->normal_int_signal_en &= 0;
+	regs->err_int_signal_en &= 0;
+	regs->timeout_ctrl = EMMC_HOST_MAX_TIMEOUT;
+}
+#endif
```
Comment (jfischer-no) [inline drivers/sdhc/emmc_host.c:81]: I guess there is no need for ugly ifdefery, Just
```c
static inline void enable_interrupts(struct emmc_data *emmc) ...
static inline void disable_interrupts(struct emmc_data *emmc) ..
```

and 
```c
if (IS_ENABLED(CONFIG_EMMC_HOST_INTR)) {
	enable_interrupts(emmc);
} else {
	disable_interrupts(emmc);
}
```

**Review by jfischer-no — COMMENTED** [review]
It is not really clear what "native driver" means in this context.

**Comment by najumon1980** [comment]
@danieldegrasse @decsny I updated the PR, could you please have a look

File: drivers/sdhc/Kconfig.emmc (hunk 7 above)
Comment (danieldegrasse) [reply drivers/sdhc/Kconfig.emmc:55]: Please remove these statements here that source `subsys/logging/Kconfig.template.log_config`

File: drivers/sdhc/Kconfig.emmc [hunk 15]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config EMMC_HOST
+	bool "EMMC driver"
```
Comment (danieldegrasse) [inline drivers/sdhc/Kconfig.emmc:5]: I am ok with the Kconfig naming here, since this driver should *theoretically* be generic across any IP implementing the SDHC host spec (if I understand the driver correctly)

However, can we add something in the Kconfig description and help string making it clear this driver is currently only validated using intel's EMMC controller, so users aren't confused?

Comment (najumon1980) [reply drivers/sdhc/Kconfig.emmc:5]: updated with validation scope.

**Review by danieldegrasse — COMMENTED** [review]
A general comment on naming here: is the driver here intended to support a host controller that complies to the [SD Host Controller Standard Specification](https://www.sdcard.org/downloads/pls/pdf?p=PartA2_SD%20Host_Controller_Simplified_Specification_Ver4.20.jpg&f=PartA2_SD%20Host_Controller_Simplified_Specification_Ver4.20.pdf&e=EN_SSA2) from the SD org, or another spec? If this implementation is intended to comply with the SD host controller specification, I think we should be naming functions with a pattern like `sd_host`, not `emmc_host`, because this aligns closer with the naming used in the specification itself.

File: drivers/sdhc/emmc_host.c [hunk 16]
```diff
@@ -0,0 +1,1334 @@
...
+		resp_type = EMMC_HOST_RESP_LEN_48;
+		break;
+	case SD_RSP_TYPE_R1b:
+		resp_type = EMMC_HOST_RESP_LEN_48B;
+		break;
+	case SD_RSP_TYPE_R2:
+		resp_type = EMMC_HOST_RESP_LEN_136;
+		break;
+
+	case SD_RSP_TYPE_R5b:
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/sdhc/emmc_host.c:622]: Currently we are not supported and as someone test the driver with SD stack then they can update this.

**Comment by najumon1980** [comment]
> It is not really clear what "native driver" means in this context.



> A general comment on naming here: is the driver here intended to support a host controller that complies to the [SD Host Controller Standard Specification](https://www.sdcard.org/downloads/pls/pdf?p=PartA2_SD%20Host_Controller_Simplified_Specification_Ver4.20.jpg&f=PartA2_SD%20Host_Controller_Simplified_Specification_Ver4.20.pdf&e=EN_SSA2) from the SD org, or another spec? If this implementation is intended to comply with the SD host controller specification, I think we should be naming functions with a pattern like `sd_host`, not `emmc_host`, because this aligns closer with the naming used in the specification itself.

Currently our scope limited eMMC and my suggestion is as someone start port this driver on their platform and test on SD card as well then then can enhance the features/scope.

**Comment by danieldegrasse** [comment]
> Currently our scope limited eMMC and my suggestion is as someone start port this driver on their platform and test on SD card as well then then can enhance the features/scope.

Understood- if the scope is limited to eMMC on the alder lake platform, then I would like to see file and Kconfig names reflect this. We can leave the C function names untouched, as those aren't really user facing, but I want it to be clear that this driver may implement the SD host controller specification, but it is only designed to function with the eMMC controller on the alder lake platform.

If we want to keep Kconfig and file names generic, then I would like this driver to be generic enough to (at a minimum) support the SD and eMMC stack. Ideally we would validate this driver on additional host controllers, but I am not aware of another one that complies fully with the SDHC spec.

**Comment by najumon1980** [comment]
> > Currently our scope limited eMMC and my suggestion is as someone start port this driver on their platform and test on SD card as well then then can enhance the features/scope.
> 
> Understood- if the scope is limited to eMMC on the alder lake platform, then I would like to see file and Kconfig names reflect this. We can leave the C function names untouched, as those aren't really user facing, but I want it to be clear that this driver may implement the SD host controller specification, but it is only designed to function with the eMMC controller on the alder lake platform.
> 
> If we want to keep Kconfig and file names generic, then I would like this driver to be generic enough to (at a minimum) support the SD and eMMC stack. Ideally we would validate this driver on additional host controllers, but I am not aware of another one that complies fully with the SDHC spec.

Agree and updated file name and Kconfig to intel specific

File: dts/bindings/sdhc/sdhc.yaml [hunk 17]
```diff
@@ -60,12 +60,8 @@ properties:
...
       The host controller supports HS400 at 1.8V
 
-  mmc-dw-8bit:
-    type: boolean
-    description: |
-      The host controller supports 8bit data width
-
-  mmc-dw-4bit:
-    type: boolean
+  bus-width:
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply dts/bindings/sdhc/sdhc.yaml:63]: Moved to original commit. Also regarding the bus width, Intel IP support both 4 and 8 bit and if someone design their board with only 4bit interface then they can update dts accordingly to limit driver settings.

**Comment by najumon1980** [comment]
@danieldegrasse I updated/respond to your comments, could you please have a look ? I would like this PR to be merged before code freeze which is this Friday

File: drivers/sdhc/intel_emmc_host.c [hunk 18]
```diff
@@ -0,0 +1,1342 @@
...
+			emmc->desc_table);
+	} else {
+		/* Setup DMA trasnfer using SDMA */
+		regs->sdma_sysaddr = (uint32_t)((uintptr_t)data->data);
+		LOG_DBG("sdma_sysaddr: %x", regs->sdma_sysaddr);
+	}
+	return 0;
+}
+
+static int32_t emmc_init_xfr(const struct device *dev, struct sdhc_data *data, bool read)
```
Comment (jhedberg) [inline drivers/sdhc/intel_emmc_host.c:449]: Why is the return type `int32_t` and not `int` here? You're assigning the return value to a variable with type `int`.

File: drivers/sdhc/intel_emmc_host.c [hunk 19]
```diff
@@ -0,0 +1,1342 @@
...
+}
+
+static int emmc_xfr(const struct device *dev, struct sdhc_command *cmd, struct sdhc_data *data,
+		    bool read)
+{
+	struct emmc_data *emmc = dev->data;
+	int ret;
+	struct emmc_cmd_config emmc_cmd;
+
+	ret = emmc_init_xfr(dev, data, read);
```
Comment (jhedberg) [inline drivers/sdhc/intel_emmc_host.c:924]: Here `ret` is of type `int`

File: drivers/sdhc/intel_emmc_host.c [hunk 20]
```diff
@@ -0,0 +1,1342 @@
...
+}
+
+static int32_t emmc_host_send_cmd(const struct device *dev, const struct emmc_cmd_config *config)
+{
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+	struct emmc_data *emmc = dev->data;
+	struct sdhc_command *sdhc_cmd = config->sdhc_cmd;
+	enum emmc_response_type resp_type = emmc_decode_resp_type(sdhc_cmd->response_type);
+	uint16_t cmd_reg;
+	int ret;
```
Comment (jhedberg) [inline drivers/sdhc/intel_emmc_host.c:667]: Here you're also mixing `int32_t` in the function prototype, but using `int ret` inside the function. There are several more cases of this in the driver. I suppose every place where you use the "0 on success or negative POSIX errno on failure" pattern you should use `int` as the type.

**Comment by najumon1980** [comment]
@danieldegrasse @jhedberg updated as per you review comment, please have a look

**Comment by najumon1980** [comment]
@decsny I update the PR as per you comments, please have a look

File: drivers/sdhc/Kconfig.intel [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+
+config INTEL_EMMC_HOST_ADMA
+	bool "EMMC host controller ADMA mode"
+	depends on INTEL_EMMC_HOST_DMA
+	help
+	  EMMC host controller ADMA mode support.
+
+config INTEL_EMMC_HOST_ADMA_DESC_SIZE
+	int "EMMC host controller ADMA Descriptor size"
+	default 32
```
[the remark this answers is not in the document]
Comment (jhedberg) [reply drivers/sdhc/Kconfig.intel:36]: > Since we use IS_ENABLED this might cause build error during compile time

It seems you're then also waisting memory, since `struct emmc_data` will always contain the `desc_table` even if ADMA is not enabled. It might be worth looking into different ways of fixing this. Some options come to mind:
- Stop using IS_ENABLED
- Refactor the code so the memory is allocated in a separate struct
- Use something like this:
```c
#ifdef CONFIG_INTEL_EMMC_HOST_ADMA_DESC_SIZE
#define ADMA_DESC_SIZE CONFIG_INTEL_EMMC_HOST_ADMA_DESC_SIZE
#else
#define ADMA_DESC_SIZE 0
#endif
```

For the record, the current code works fine, so this could also be done as a follow-up commit (especially considering that time is running low on getting a new driver merged before the feature freeze)

[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/sdhc/Kconfig.intel:37]: updated as per your suggestion

**Review by jhedberg — DISMISSED** [review]

**Comment by decsny** [comment]
@najumon1980 

please see https://github.com/zephyrproject-rtos/zephyr/pull/59886 about bus width property, your PR conflicts conceptually with what is in tree currently

File: subsys/sd/sd.c [hunk 22]
```diff
@@ -117,17 +117,32 @@ static int sd_common_init(struct sd_card *card)
...
+	} else if (host_props->host_caps.vol_300_support) {
+		LOG_DBG("Host controller support 3.0V max");
+		voltage = SD_VOL_3_0_V;
+	} else {
+		LOG_DBG("Host controller support 1.8V max");
+		voltage = SD_VOL_1_8_V;
+	}
+
+	/* Set to maximum voltage support by Host controller */
+	card->card_voltage = voltage;
```
[the remark this answers is not in the document]
Comment (danieldegrasse) [reply subsys/sd/sd.c:144]: Not sure I follow- The issue I am highlighting is that `card->card_voltage` is being set twice in this function, and it only needs to be set once (on line 163). The initial logic here set `card_voltage` once on line 163, and your changes introduce a second assignment to it on line 144. Is there a reason we need this assignment?

**Comment by danieldegrasse** [comment]
@najumon1980- just to be clear- please address my comment about the duplicate assignment to `card->card_voltage` as well as the `bus-width` DT property. Other than that I am ok to merge this PR.

File: dts/bindings/sdhc/sdhc.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (najumon1980) [reply dts/bindings/sdhc/sdhc.yaml]: agree and updated the code

File: subsys/sd/sd.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (najumon1980) [reply subsys/sd/sd.c]: removed the duplicate code

**Review by jhedberg — APPROVED** [review]

**Comment by najumon1980** [comment]
> @najumon1980- just to be clear- please address my comment about the duplicate assignment to `card->card_voltage` as well as the `bus-width` DT property. Other than that I am ok to merge this PR.

@danieldegrasse I updated the PR and pls have a look

**Review by danieldegrasse — APPROVED** [review]
